ntp-udp/src/raw_socket.rs

```diff
@@ -0,0 +1,386 @@
+/// This file contains safe wrappers for the socket-related system calls
+/// needed to implement the UdpSocket in socket.rs
+///
+/// Since the safety of a rust unsafe block depends not only on its
+/// contents, but also the context within which it is called, the code
+/// here is split up in submodules that are individually as small as
+/// possible while still having each a fully safe API interface. This
+/// should reduce the amount of context which needs to be considered
+/// when reasoning about safety, significantly simplifying the checking
+/// of this code.
+///
+/// All unsafe blocks are preceded with a comment explaining why that
+/// specific unsafe code should be safe within the context in which it
+/// is used.
+
+mod exceptional_condition_fd {
+    use std::os::unix::prelude::{AsRawFd, RawFd};
+
+    use tokio::io::unix::AsyncFd;
+
+    use super::cerr;
+
+    pub(crate) fn exceptional_condition_fd(
+        socket_of_interest: &std::net::UdpSocket,
+    ) -> std::io::Result<AsyncFd<RawFd>> {
```

Comment:
maybe add some docs here on what an this is (with some links to linux docs). It also would be really nice if we could upstream this, maybe also link the tokio issue about it that we have somewhere but that I can't find right now

Reply:
added addition comment